osh/cmd_eval.py

```diff
@@ -1433,6 +1433,21 @@ def _Dispatch(self, node, cmd_st):
   def RunPendingTraps(self):
     # type: () -> None
 
+    # Round up traps for signals fired in C++
+    # XXX: The SignalState_PendingTraps()/SignalState_ClearPendingTraps() API
+    # from pyos here feels a little clumsy, but it's not clear there's a good
+    # way around it. We need access to this CommandEvaluator in order to execute
+    # the trap code, but the type isn't accessible from the hand-written C++
+    pending = pyos.SignalState_PendingTraps()
+    if len(pending):
+      pending = list(pending)
+      pyos.SignalState_ClearPendingTraps()
+      for sig_num, handler in pending:
+          # TODO: Right now we have to invoke __call__() explicitly instead of
+          # using it as sugar. Perhaps we should teach mycpp to translate
+          # __call__() into operator() in C++
+          handler.__call__(sig_num, None)
```

Comment:
OK I see the key difficulty here ...

Though I'm a little surprised that the `__call__` translates to C++?  I didn't run it but it looks like the CI passes?

Reply:
Yeah, I was a little surprised by that too. It builds and passes `spec-cpp/builtins-trap` and a few manual tests, though

Reply:
Another option could be to expose more of `_TrapHandler` to the hand-written C++. Probably the same type of work, though. Just less of it since `_TrapHandler` is much simpler than `CommandEvaluator`